text/0000-overload-logic-operators.md

```diff
@@ -0,0 +1,211 @@
+- Feature Name: overload-logic-operators
+- Start Date: 2019-07-04
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+This feature would allow for the two short circuit operators `||` and `&&` to be overloadable by
+users-of-rust and for the standard library (probably in core/alloc) to implement such an overload for
+the `Option<T>` and `Result<T, E>` types for `||` and for the `Option<T>` type `&&` but not for
+`Result<T, E>`.
+
+# Motivation
+[motivation]: #motivation
+
+This idea was original floated as a way to clear up the differences between `.or(...)`, `.or_with(|| ...)`, `.and(...)`, `.and_with(|| ...)`, `.unwrap_or(...)`, and `.unwrap_or_with(|| ...)`. Not only was the requirement to remember that there were additional methods that are supposed to be used when you don't want to compute the value before the check (short circuiting). There was also a concern about the overhead of the additional closure.
+
+This proposal is mostly about reducing the mental strain when chaining `Option<T>`'s and `Result<T, E>`'s. But has a very nice side effect of allowing users-of-rust the ability to overload these operators.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+This proposal starts with an enum definition and trait definitions for each of the operators:
```

Comment:
I don't think this is the appropriate start for a guide-level explanation.  I think this section should look substantially more like the `?` description in the book: describe a common need, describe how it can be done manually with `if let`, then describe how it's handled using the `||`/`&&` operators to be more concise.  I think this section should also emphasize the parallel with booleans -- how `foo() && bar()` is `if foo() { true } else { bar() }` and how that's the same pattern in the `if let`s seen here.